### PR TITLE
Support range sortKey feature

### PR DESCRIPTION
**Pull request by javeme**
implement #271

Change-Id: Ib618c45bc6169a0fb7632dbbf0fb29b7ee0a08e3

### CODE REVIEW

File: hugegraph-test/src/main/resources/hugegraph.properties [hunk 1]
```diff
@@ -1,7 +1,10 @@
 gremlin.graph=com.baidu.hugegraph.HugeFactory
 
-backend=${backend}
-serializer=${serializer}
+#backend=${backend}
```
Comment (Linary) [inline hugegraph-test/src/main/resources/hugegraph.properties]: keep origin

File: hugegraph-test/src/main/java/com/baidu/hugegraph/unit/UnitTestSuite.java [hunk 2]
```diff
@@ -36,6 +37,7 @@
     RamCacheTest.class,
     CacheManagerTest.class,
 
+    VersionTest.class,
```
Comment (Linary) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/unit/UnitTestSuite.java]: remove

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/ConditionQuery.java [hunk 3]
```diff
@@ -277,11 +277,11 @@ public void resetUserpropConditions() {
 
     /**
      * This method is only used for secondary index scenario,
-     * relation must be EQ
+     * its relation must be EQ
      * @param fields the user property fields
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/ConditionQuery.java]: align with 282

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdPrefixQuery.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    public boolean empty() {
+        return false;
+    }
+
+    @Override
+    public boolean test(HugeElement element) {
+        byte[] elem = element.id().asBytes();
+        int cmp = Bytes.compare(elem, this.start.asBytes());
+        boolean matchedStart = this.inclusiveStart ? cmp >= 0 : cmp > 0;
+        boolean matchedPrefix = Bytes.prefixWith(elem, this.prefix.asBytes());
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdPrefixQuery.java]: A little performance boost：
 if (!matchedStart) {
    return false
}
...

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdPrefixQuery.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        return matchedStart && matchedPrefix;
+    }
+
+    @Override
+    public IdPrefixQuery copy() {
+        IdPrefixQuery query = (IdPrefixQuery) super.copy();
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdPrefixQuery.java]: just return (IdPrefixQuery) super.copy();

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdRangeQuery.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    public boolean test(HugeElement element) {
+        int cmp1 = Bytes.compare(element.id().asBytes(), this.start.asBytes());
+        int cmp2 = Bytes.compare(element.id().asBytes(), this.end.asBytes());
+        return (this.inclusiveStart ? cmp1 >= 0 : cmp1 > 0) &&
+               (this.inclusiveEnd ? cmp2 <= 0 : cmp1 < 0);
+    }
+
+    @Override
+    public IdRangeQuery copy() {
+        IdRangeQuery query = (IdRangeQuery) super.copy();
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdRangeQuery.java]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdRangeQuery.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    @Override
+    public boolean empty() {
+        return false;
+    }
+
+    @Override
+    public boolean test(HugeElement element) {
+        int cmp1 = Bytes.compare(element.id().asBytes(), this.start.asBytes());
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdRangeQuery.java]: seems can also be optimized

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/ConditionQuery.java (hunk 3 above)
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/ConditionQuery.java]: indentation

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdRangeQuery.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        int cmp2 = Bytes.compare(element.id().asBytes(), this.end.asBytes());
+        return (this.inclusiveStart ? cmp1 >= 0 : cmp1 > 0) &&
+               (this.inclusiveEnd ? cmp2 <= 0 : cmp1 < 0);
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdRangeQuery.java]: (this.inclusiveEnd ? cmp2 <= 0 : cmp2 < 0);

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/PropertyKey.java [hunk 9]
```diff
@@ -136,6 +165,17 @@ public HugeType type() {
         return valid;
     }
 
+    public <V> Object serialValue(V value) {
+        V validValue = this.validValue(value);
+        E.checkArgument(value != null,
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/PropertyKey.java]: validValue

File: hugegraph-example/src/main/java/com/baidu/hugegraph/example/Example1.java [hunk 10]
```diff
@@ -407,8 +407,10 @@ public static void testRemove(final HugeGraph graph) {
         assert vertices.size() == 1;
         Vertex james = vertices.get(0);
         Vertex book6 = graph.addVertex(T.label, "book", "name", "java-6");
-        james.addEdge("look", book6, "time", "2017-5-3");
+        james.addEdge("look", book6, "timestamp", "2017-5-2 12:00:08.0");
+        james.addEdge("look", book6, "timestamp", "2017-5-3 12:00:08.0");
         graph.tx().commit();
+System.exit(0);
```
Comment (zhoney) [inline hugegraph-example/src/main/java/com/baidu/hugegraph/example/Example1.java]: delete

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdPrefixQuery.java (hunk 4 above)
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdPrefixQuery.java]: ok

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdPrefixQuery.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        this(resultType, null, prefix, true, prefix);
+    }
+
+    public IdPrefixQuery(Query originQuery, Id prefix) {
+        this(originQuery.resultType(), originQuery, prefix, true, prefix);
+    }
+
+    public IdPrefixQuery(Query originQuery,
+                         Id start, boolean inclusiveStart, Id prefix) {
+        this(originQuery.resultType(), originQuery, start, true, prefix);
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/IdPrefixQuery.java]: true -> inclusiveStart

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 12]
```diff
@@ -971,11 +970,34 @@ protected Query optimizeQuery(ConditionQuery query) {
...
+                    PropertyKey pk = this.graph().propertyKey((Id) r.key());
+                    r.serialValue(pk.serialValue(r.value()));
+                }
+                // Convert to sysprop condition {SORT_VALUES=value}
+                if (query.hasRangeCondition()) {
+                    for (Condition condition : query.userpropConditions()) {
+                        assert condition instanceof Condition.Relation;
+                        Condition.Relation r = (Condition.Relation) condition;
+                        Condition.Relation sys = new Condition.SyspropRelation(
+                                HugeKeys.SORT_VALUES,
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java]: Adjust the alignment

**Pull request merged**